src/python/pants/bsp/rules.py

```diff
@@ -71,9 +86,72 @@ async def bsp_workspace_build_targets(
     )
 
 
+# -----------------------------------------------------------------------------------------------
+# Build Target Sources Request
+# See https://build-server-protocol.github.io/docs/specification.html#build-target-sources-request
+# -----------------------------------------------------------------------------------------------
+
+
+@dataclass(frozen=True)
+class MaterializeBuildTargetSourcesRequest:
+    bsp_target_id: BuildTargetIdentifier
+
+
+@dataclass(frozen=True)
+class MaterializeBuildTargetSourcesResult:
+    sources_item: SourcesItem
+
+
+@rule
+async def materialize_bsp_build_target_sources(
+    request: MaterializeBuildTargetSourcesRequest,
+    build_root: BuildRoot,
+) -> MaterializeBuildTargetSourcesResult:
+    uri = request.bsp_target_id.uri
+    if not uri.startswith("pants:"):
+        raise ValueError(f"Unknown URI for Pants BSP: {uri}")
```

Comment:
Ditto on eagerly validating this w/ `__post_init__`

Reply:
And probably exposing `def address_input` as a property that does the URI stripping.